### PR TITLE
Density3

### PR DESCRIPTION
**Pull request by sombrereau**
## Description

Resolve discrepancy between density3 and density https://github.com/terrapower/armi/issues/865 for open source ARMI materials.

This includes:
- creating a subclass `SimpleSolid` material to help define very simple materials. 
- adding to the base class of the materials unit test to verify the `density3` method produces a non-zero result
- updating the framework materials and the unit tests so they all pass
---

## Checklist

- [X] This PR has only one purpose or idea.
- [X] Tests have been added/updated to verify that the new/changed code works.

- [X] The code style follows [good practices](https://terrapower.github.io/armi/developer/standards_and_practices.html).
- [X] The commit message(s) follow [good practices](https://terrapower.github.io/armi/developer/tooling.html).

- [X] The [release notes](https://terrapower.github.io/armi/release/index.html) (location `doc/release/0.X.rst`) are up-to-date with any bug fixes or new features.
- [X] The documentation is still up-to-date in the `doc` folder.
- [X] The dependencies are still up-to-date in `setup.py`.

### CODE REVIEW

**Comment by albeanth** [comment]
Hi @sombrereau ! @john-science requested that I give this a review as well. 

I'm still making my way through your changes but have a thought. What if, instead of a `_density()` method on each material, we add a constructor to each material and directly set `self.p.refDens` and `self.refTempK`? I think it might be clearer as some materials have the `_density()` method and some don't. It also might be confusing relative to `armi.material.density`. Just a thought. 

One more: What do you think about making `self.refTempK` a param? I could see it being useful to know on a DB load what the reference temp refDens correlates to.

**Comment by albeanth** [comment]
A few examples of how the reference density is being set for different materials. I think it would be great to have consistency - maybe through the constructor of each material? I'm certainly open to other ideas; I just advocate for consistency and clarity. 

https://github.com/terrapower/armi/blob/4d925b522141be7352276de4b5c80855136970ec/armi/materials/b4c.py#L22

https://github.com/terrapower/armi/blob/4d925b522141be7352276de4b5c80855136970ec/armi/materials/concrete.py#L42-L43

https://github.com/terrapower/armi/blob/4d925b522141be7352276de4b5c80855136970ec/armi/materials/ht9.py#L60

https://github.com/terrapower/armi/blob/afd2ead0db0328b16af9319d3aab76255e167a7a/armi/materials/californium.py#L33-L37

**Review by albeanth — COMMENTED** [review]
@sombrereau feel free to take these comments as you see fit. As we briefly chatted, `armi.materials` have some inconsistencies that could benefit from some attention. Giving this the attention it deserves is likely out of scope for this PR - a few of the comments I have given may also be out of scope. Feel free to kick the can oh what you'd like. 

@john-science there may be some benefit in holding some conversations/discussions on what we'd like to change with `armi.materials`.

**Comment by john-science** [comment]
> A few examples of how the reference density is being set for different materials. I think it would be great to have consistency - maybe through the constructor of each material? I'm certainly open to other ideas; I just advocate for consistency and clarity.

Tony, obviously, I 100% agree.

But I think that would be a bigger change to the system than this PR aims for.  Tommy identified a real bug, where a lot of materials are reporting a density3 of 0.  Which... makes me unhappy.

I like the idea though.  Can you propose that in our umbrella "materials make us sad" discussion?    https://github.com/terrapower/armi/discussions/767

File: armi/materials/mgO.py [hunk 1]
```diff
@@ -27,23 +27,18 @@ class MgO(Material):
         "linear expansion percent": ((273.15, 1273.15), "K"),
     }
 
+    def __init__(self):
+        Material.__init__(self)
+        """same reference as linear expansion. Table II.
+        Reference density is from Wolfram Alpha At STP (273 K)"""
+
+        self.p.refDens = 3.58
```
[the remark this answers is not in the document]
Comment (john-science) [reply armi/materials/mgO.py:35]: That's a great idea!

But, just my opinion, that should be a second PR.  Tommy is just trying to fix a bug here.

And, again, please don't get me wrong. ARMI materials need work. I'm happy to consider ALL kinds of changes.  Maybe you could post that to our ARMI Materials discussion?  https://github.com/terrapower/armi/discussions/767

**Comment by albeanth** [comment]
> Can you propose that in our umbrella "materials make us sad" discussion? 

@john-science definitely. I think I missed the fact that Tommy's PR was to simply fix materials with a density3 of 0.0 - I thought he was aiming to address #865 in full.

I realize that many of my comments may result in scope creep if they were actually addressed... Those that do should 100% be punted to future work. We can iterate on which ones fit that bill.

File: armi/materials/uraniumOxide.py [hunk 2]
```diff
@@ -162,7 +164,7 @@ def meltingPoint(self):
         """
         return self.__meltingPoint
 
-    def density(self, Tk: float = None, Tc: float = None) -> float:
+    def _density(self, Tk: float = None, Tc: float = None) -> float:
```
[the remark this answers is not in the document]
Comment (john-science) [reply armi/materials/uraniumOxide.py:167]: Yeah, that's going to be a big push at some point soon.

BUT let's not force that separate ticket on Tommy for this PR.  He's just fixing a bug.

I think that other ticket will force big changes on the simulation results of downstream ARMI plugins and apps.  Which maybe is what we want. But it will definitely require a few of it's own PRs, so that we can _very carefully_ test HOW downstream modeling results change.

It's import, for sure. But we also have a separate ticket for that change.

**Comment by john-science** [comment]
@sombrereau Okay, so the idea here is that we have a bunch of 3D densities incorrectly reported as 3D?  BUT... instead of just changing the name from `density()` to `density3()` for all these materials, you create the `_density()` method, because we can use that for your new linear expansion method?

Sorry, I'm not _trying_ to be obtuse.  Just trying to figure it out.

Why didn't we just change all the incorrect `density` to `density3`?

File: armi/materials/thoriumOxide.py [hunk 3]
```diff
@@ -74,9 +79,8 @@ def setDefaultMassFracs(self):
         self.setMassFrac("TH232", 0.8788)
         self.setMassFrac("O16", 0.1212)
 
-    def density(self, Tk=None, Tc=None):
-        """g/cc from IAEA TE 1450"""
-        return 10.00 * self.getTD()
+    def density3(self, Tk: float = None, Tc: float = None) -> float:
+        return Material.density(self, Tk=Tk, Tc=Tc)
```
[the remark this answers is not in the document]
Comment (sombrereau) [reply armi/materials/thoriumOxide.py:83]: nice! good catch!

File: armi/utils/densityTools.py [hunk 4]
```diff
@@ -376,6 +376,30 @@ def getChemicals(nuclideInventory):
...
+def calculateLinearExpansion(density1, density2):
+    """
+    calculates linear expansion based on two densities
+
+    See Also
+    --------
+    materials.material.linearExpansionPercent
+    utils.densityTools.calculateLinearExpansionPercent
+    """
+    return (density1 / density2) ** (1.0 / 3.0) - 1
```
[the remark this answers is not in the document]
Comment (sombrereau) [reply armi/utils/densityTools.py:388]: this is a good point, I only really need it in the Simple Solid Materials, I'll just move what I need there.

**Review by albeanth — APPROVED** [review]

**Review by john-science — APPROVED** [review]

File: armi/materials/tests/test_materials.py [hunk 5]
```diff
@@ -43,6 +45,11 @@ def test_isPicklable(self):
             self.mat.thermalConductivity(500), mat.thermalConductivity(500)
         )
 
+    def test_density3(self):
+        """Test that all materials produce a zero density from density3"""
```
Comment (keckler) [inline armi/materials/tests/test_materials.py:49]: @sombrereau 

The docstring here does not appear to align with what is actually being tested.

Can you clarify which way is meant to be true, i.e. if `density3()` is supposed to be zero or not?

File: armi/materials/material.py [hunk 6]
```diff
@@ -690,7 +690,71 @@ def density3(self, Tk=None, Tc=None):
...
+    This function assumed the density is defined on the _density method and
+    this base class keeps density, density3 and linearExpansion all in sync
+
+    class SimpleMaterial(SimpleSolid):
+
+        def _density(self, Tk=None, Tc=None):
+            "
+            density that preserves mass when thermally expanded in 3D.
+            "
+            ...
```
Comment (keckler) [inline armi/materials/material.py:708]: @sombrereau Can you clarify what is meant here with the `_density()` method?

Based on the docstring, it appears that it should be necessary to define a `_density()` method, but looking around the code, I don't ever see one.

Am I reading this incorrectly? Or is this docstring out of date?

Comment (john-science) [reply armi/materials/material.py:708]: Out of date.

I am working on a PR that moves:

* `Material.density()` to `Material.pseudoDensity()`
* `Material.density3()` to `Material.density()`

And in that branch, I have already fixed this docstring.

(Also, spoilers, I found a bunch of bugs while making that branch.  Joy!)
